src/proof/uf_proof.cpp

```diff
@@ -649,11 +649,13 @@ void LFSCUFProof::printOwnedTermAsType(Expr term,
     os << "(apply _ _ ";
   }
   os << func << " ";
+  Assert(func.getType().isFunction());
+  FunctionType funcType = static_cast<FunctionType>(func.getType());
```

Comment:
In another PR you eventually preferred to use a node representation instead of an expr representation of types.
This is also possible here, by using `fromExpr` function of `Node` on `term`, and then use `getOperator().getType().getArgTypes()`. It could be better than a static cast, though I am not sure. It would slightly diminish the usage of Expr but not entirely. Your call, and maybe @4tXJ7f has an opinion here?

Reply:
Since this is a small-scope local variable I'm not immediately worried about using `Expr`. That being said, I defer to Andres. I don't actually know how these structures work.

Reply:
I'll just change it. Thanks for the suggestion!

Reply:
Internally, we should never use Expr, always Node. Expr is/was the old API layer, and should not be used in the solver internals.

Reply:
Yes, we should not use `Expr`-layer classes internally as @aniemetz said. I would also like to add that this code actually results in undefined behavior: You can't cast an object of type `Type` to `FunctionType`. It kind of works in practice because `FunctionType` has no additional data members nor virtual methods (see also #3420) but is dangerous.

Reply:
Good to know. I know less about class hierarchies in C++ than I would like to.

This is gone now.